In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("./Resources/lending_data.csv")
lending_data_df = pd.read_csv(file_path)

'''A value of 0 in the “loan_status” column means that the loan is healthy. 
A value of 1 means that the loan has a high risk of defaulting.'''

# Review the DataFrame
lending_data_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels/target
y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.drop(columns=['loan_status'])

In [4]:
#count how many rows 
len(X)

77536

In [5]:
# Review the y variable Series
display(y.head())
display(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1,
                                                    stratify=y)
X_train.shape

(58152, 7)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs',
                                max_iter=200,
                                random_state=1)


# Fit the model using training data
fitted_model = logistic_regression_model.fit(X_train, y_train)
logistic_regression_model

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(X_test)

#look at training predictions
training_predictions = fitted_model.predict(X_train)


print(f"Training Data Score: {logistic_regression_model.score(X_train, y_train)}")
print(f"Testing Data Score: {logistic_regression_model.score(X_test, y_test)}")


results_df = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test}).tail(10)
results_df

Training Data Score: 0.9926743706149401
Testing Data Score: 0.993912505158894


Prediction  Actual
76038           1       1
19119           0       0
45193           0       0
25211           0       0
61697           0       0
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
from sklearn.metrics import accuracy_score
acc_score_test = accuracy_score(y_test, testing_predictions)
acc_score_test

0.993912505158894

In [11]:
acc_score_train = accuracy_score(y_train, training_predictions)
acc_score_train

0.9926743706149401

In [12]:
#generate a confusion matrix for the model (training matrix)
cm_train = confusion_matrix(y_train, training_predictions)
print(cm_train)

cm_train_df = pd.DataFrame(
    cm_train, index=["actual_0_healthy_loan", "actual_1_high_risk_loan"], columns=["predicted_0_healthy_loan", "predicted_1_high_risk_loan"]
)
cm_train_df

[[55967   310]
 [  116  1759]]


predicted_0_healthy_loan  predicted_1_high_risk_loan
actual_0_healthy_loan                       55967                         310
actual_1_high_risk_loan                       116                        1759

In [13]:
# Generate a confusion matrix for the model (testing data)
cm_test = confusion_matrix(y_test, testing_predictions)
print(cm_test)

cm_test_df = pd.DataFrame(
    cm_test, index=["actual_0_healthy_loan", "actual_1_high_risk_loan"], columns=["predicted_0_healthy_loan", "predicted_1_high_risk_loan"]
)
cm_test_df


[[18673    86]
 [   32   593]]


predicted_0_healthy_loan  predicted_1_high_risk_loan
actual_0_healthy_loan                       18673                          86
actual_1_high_risk_loan                        32                         593

In [14]:
# Print the classification report for the model (training classification report)
target_names = ["healthy_loan", "high_risk_loan"]
class_report_train = (classification_report(y_train, training_predictions, target_names=target_names))


In [15]:
# Print the classification report for the model (testing classification report)
target_names = ["healthy_loan", "high_risk_loan"]
class_report_test = (classification_report(y_test, testing_predictions, target_names=target_names))


In [16]:
#display results for train 
print("Confusion Matrix - Train")
display(cm_train_df)
print(f"Accuracy Score : {acc_score_train}")
print("Classification Report")
print(class_report_train)

Confusion Matrix - Train


predicted_0_healthy_loan  predicted_1_high_risk_loan
actual_0_healthy_loan                       55967                         310
actual_1_high_risk_loan                       116                        1759

Accuracy Score : 0.9926743706149401
Classification Report
                precision    recall  f1-score   support

  healthy_loan       1.00      0.99      1.00     56277
high_risk_loan       0.85      0.94      0.89      1875

      accuracy                           0.99     58152
     macro avg       0.92      0.97      0.94     58152
  weighted avg       0.99      0.99      0.99     58152



In [17]:
#display results for test
print("Confusion Matrix - Test")
display(cm_test_df)
print(f"Accuracy Score : {acc_score_test}")
print("Classification Report")
print(class_report_test)

Confusion Matrix - Test


predicted_0_healthy_loan  predicted_1_high_risk_loan
actual_0_healthy_loan                       18673                          86
actual_1_high_risk_loan                        32                         593

Accuracy Score : 0.993912505158894
Classification Report
                precision    recall  f1-score   support

  healthy_loan       1.00      1.00      1.00     18759
high_risk_loan       0.87      0.95      0.91       625

      accuracy                           0.99     19384
     macro avg       0.94      0.97      0.95     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model predicts a healthy loan with 100% precision and a high-risk loan at 87% precision. 

---